libcontainer/container_linux.go

```diff
@@ -2358,3 +2378,120 @@ func requiresRootOrMappingTool(c *configs.Config) bool {
 	}
 	return !reflect.DeepEqual(c.GidMappings, gidMap)
 }
+
+// setCPUAffinity adjusts the CPU affinity if necessary based on heuristics.
+func setCPUAffinity(fsys fs.FS, cpusetList string, setFn func(uint32)) error {
+	schedRuntimeData, err := fs.ReadFile(fsys, "proc/sys/kernel/sched_rt_runtime_us")
+	if err != nil || len(schedRuntimeData) == 0 {
+		fmt.Println(err)
+		return nil
+	}
+	schedRuntime, err := strconv.ParseInt(string(bytes.TrimSpace(schedRuntimeData)), 10, 64)
+	if err != nil {
+		return err
+	}
+
+	// When set to -1, it gives zero room for kernel to switch in kernel space on
+	// a CPU where a real-time application is running on under scheduler real-time
+	// policies and do not make any context switch.
+	// With this setting value and because the kernel randomly dispatch runc process
+	// on one of CPUs, runc can get stuck under some conditions during cgroup
+	// transition when many tasks within a cgroup cpuset are running under a scheduler
+	// real-time policy and on isolated CPUs.
+	// The call to getEligibleCPU below returns a predictable CPU affinity to circumvent
+	// this kernel random dispatching when isolated CPUs are present in the cpuset.
+	if schedRuntime != -1 {
+		return nil
+	}
+
+	// get the isolated CPU list
+	isolatedData, err := fs.ReadFile(fsys, "sys/devices/system/cpu/isolated")
+	// The above file may not be available in some environment
+	// due to /sys not being mounted, if not present we don't
+	// try to set CPU affinity and ignore the error.
+	// When an empty set is returned, the data length is equal
+	// to 1 (newline char), when set its length is greater than 1
+	// which means we may need to adjust CPU affinity shortly.
+	if err != nil || len(isolatedData) == 0 {
+		return nil
+	}
+
+	cpu, eligible, err := getEligibleCPU(cpusetList, string(bytes.TrimSpace(isolatedData)))
+	if err != nil {
+		return fmt.Errorf("getting eligible cpu: %w", err)
+	} else if eligible && setFn != nil {
+		setFn(uint32(cpu))
+	}
+
+	return nil
+}
+
+// getEligibleCPU returns the first eligible CPU for CPU affinity before
+// entering in a cgroup cpuset:
+//   - when there is not cpuset cores: no eligible CPU
+//   - when there is not isolated cores: no eligible CPU
+//   - when cpuset cores are all isolated cores: return the first CPU of the cpuset
+//   - when cpuset cores are mixed between housekeeping/isolated cores: return the
```

Comment:
There is another situation: when all cpuset cores are not isolated cores.
This function should return -1, false, nil.

Reply:
And add a test case to cover this situation, thanks.

Reply:
This is covered by `isolatedCores > 0` https://github.com/opencontainers/runc/pull/3923/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2496 which is greater than 0 when there at least one core present in both sets.

And here the corresponding test case https://github.com/opencontainers/runc/pull/3923/files#diff-a632fc3b003cd903187e1091e7699141c62c87887c4a5e5ab17f99cb3de427c0R392

Reply:
I updated the function comment as a covered case

Reply:
To simplify this function, I have another idea. Could you consider break the for loop and return when isolatedCores>0 && affinityCore>-1?

Reply:
> Yes, you covered it.

You have covered these two cases in one case:
1. cpuset: 0-1, isolated: 10-11
2. cpuset: 10-11, isolated: 0-1

Reply:
> Yes, you covered it. My original thought is that we should set affinityCore to -1, if isolatedCores==0, or combine this two returns to one? Keep affinityCore, remove isolatedCores?

Fixed

> To simplify this function, I have another idea. Could you consider break the for loop and return when isolatedCores>0 && affinityCore>-1?

As we need to handle the mixed case, we can only return when mixed case is detected and isolatedCores > 0, it's covered here https://github.com/opencontainers/runc/pull/3923/files#diff-1cdede3fe918430ca0b7e4056abc49abca8a2925c6770ec1951befa6091f7a94R2483

Globally, the loop has been a bit refactored to be less verbose. Thanks !

Reply:
> You have covered these two cases in one case:

    cpuset: 0-1, isolated: 10-11
    cpuset: 10-11, isolated: 0-1

Fixed too